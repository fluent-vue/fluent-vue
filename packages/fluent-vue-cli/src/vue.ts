import type { SFCBlock, SFCDescriptor } from '@vue/compiler-sfc'
import type { MessagesWithLocale } from './types'

import { parse } from '@vue/compiler-sfc'
import { parse as parseDom, compile as compileDom } from '@vue/compiler-dom'
import { merge as mergeFtl, getMessages as getFtlMessages } from './ftl'

export function getMessages (source: string): MessagesWithLocale[] {
  const descriptor = getDescriptor(source)
  descriptor.customBlocks.sort((a, b) => {
    return a.loc.start.offset - b.loc.start.offset
  })
  return extractFromCustomBlocks(descriptor.customBlocks)
}

export function merge (source: string, locale: string, messages: Record<string, string>): string {
  const parseResult = getDescriptor(source)

  let fluentBlock = parseResult.customBlocks
    .find(block => block.type === 'fluent' && block.attrs.locale === locale)

  if (fluentBlock != null) {
    const newBlockContent = mergeFtl(fluentBlock.content, messages)
    fluentBlock.content = newBlockContent
  } else {
    fluentBlock = {
      type: 'fluent',
      attrs: {
        locale: locale
      },
      content: mergeFtl('', messages)
    } as unknown as SFCBlock
  }

  // Write back
  const blocks = getBlocks(parseResult)
  return buildContent(fluentBlock, parseResult.source, blocks)
}

const decodeRE = /&(gt|lt|amp|apos|quot);/g
const decodeMap = {
  gt: '>',
  lt: '<',
  amp: '&',
  apos: "'",
  quot: '"'
}

function getDescriptor (source: string): SFCDescriptor {
  return parse(source, {
    sourceMap: false,
    ignoreEmpty: false,
    pad: false,
    compiler: {
      parse (template, options) {
        return parseDom(template, {
          ...options,
          decodeEntities: (rawText) => rawText.replace(decodeRE, (_, p1: keyof typeof decodeMap) => decodeMap[p1])
        })
      },
      compile: compileDom
    }
  }).descriptor
}

function extractFromCustomBlocks (blocks: SFCBlock[]): MessagesWithLocale[] {
  return blocks.map(block => {
    if (block.type !== 'fluent') {
      return undefined
    }

    const locale = block.attrs.locale
    if (locale == null || typeof locale !== 'string') {
      throw new Error('fluent custom block does not have locale specified')
    }

    return {
      locale,
      messages: getFtlMessages(block.content)
    }
  }).filter(Boolean) as MessagesWithLocale[]
}

function getBlocks (descriptor: SFCDescriptor): SFCBlock[] {
  const { template, script, styles, customBlocks } = descriptor

  const blocks: SFCBlock[] = [...styles, ...customBlocks]
  ;(template != null) && blocks.push(template as SFCBlock)
  ;(script != null) && blocks.push(script as SFCBlock)
  blocks.sort((a, b) => {
    return a.loc.start.offset - b.loc.start.offset
  })
  return blocks
}

function buildContent (blockToAdd: SFCBlock, raw: string, blocks: SFCBlock[]): string {
  const EOL = /\r?\n|\r/g.exec(raw)?.[0]
  if (EOL == null) {
    // Empty input
    return raw
  }

  let offset = 0
  let inserted = false
  let contents: string[] = []
  let fluentOffset = -1

  contents = blocks.reduce((contents, block, i) => {
    if (block.type === 'fluent' && block.attrs.locale === blockToAdd.attrs.locale) {
      contents = contents.concat(raw.slice(offset, block.loc.start.offset))
      contents = contents.concat(`${EOL}${blockToAdd.content}`)
      offset = block.loc.end.offset
      inserted = true
    } else {
      contents = contents.concat(raw.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset

      if (block.type === 'fluent') {
        fluentOffset = contents.join('').length + '</fluent>'.length + EOL.length
      }
    }
    return contents
  }, contents)

  contents = contents.concat(raw.slice(offset, raw.length))

  let source = contents.join('')

  if (!inserted) {
    const content = `${EOL}<fluent locale="${blockToAdd.attrs.locale as string}">${EOL}${blockToAdd.content}</fluent>${EOL}`

    if (fluentOffset !== -1) {
      source = source.slice(0, fluentOffset) + content + source.slice(fluentOffset)
    } else {
      source = source + content
    }
  }

  return source
}
