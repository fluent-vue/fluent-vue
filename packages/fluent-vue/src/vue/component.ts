import type { VueComponent } from '../types/typesCompat'

import { defineComponent, h, getCurrentInstance, inject, computed } from 'vue-demi'
import { camelize } from '../util/camelize'
import { getContext } from '../composition'
import { RootContextSymbol } from '../symbols'
import { assert } from '../util/warn'

function getParentWithFluent(
  instance: VueComponent | null | undefined
): VueComponent | null | undefined {
  const parent = instance?.$parent
  const target = parent?.$options

  if (target != null && target.fluent == null) {
    return getParentWithFluent(parent)
  }

  return parent
}

export default defineComponent({
  name: 'i18n',
  props: {
    path: { type: String, required: true },
    tag: { type: String, default: 'span' },
    args: { type: Object, default: () => ({}) },
  },
  setup(props, { slots, attrs }) {
    const rootContext = inject(RootContextSymbol)
    assert(rootContext != null, 'i18n component used without instaling plugin')
    const instance = getCurrentInstance()
    const parent = getParentWithFluent(instance?.proxy)
    const fluent = getContext(rootContext, parent)

    const fluentParams = computed(() =>
      Object.assign(
        {},
        props.args,
        ...Object.keys(slots).map((key) => ({
          [key]: `\uFFFF\uFFFE${key}\uFFFF`,
        }))
      )
    )

    const translation = computed(() => {
      return fluent.formatWithAttrs(props.path, fluentParams.value)
    })

    const camelizedAttrs = computed(() =>
      Object.fromEntries(
        Object.entries(translation.value.attributes).map(([key, value]) => [camelize(key), value])
      )
    )

    return () =>
      h(
        props.tag,
        {
          ...attrs,
        },
        translation.value.value
          .split('\uFFFF')
          .map((text) =>
            text.startsWith('\uFFFE')
              ? slots[text.replace('\uFFFE', '')]?.(camelizedAttrs.value)
              : text
          )
      )
  },
})
