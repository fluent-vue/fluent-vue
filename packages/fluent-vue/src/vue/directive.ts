import { warn } from '../util/warn'
import { TranslationContext } from '../TranslationContext'
import { getContext } from '../composition'

// This part is from fluent-dom library
const LOCALIZABLE_ATTRIBUTES = {
  global: ['title', 'aria-label', 'aria-valuetext', 'aria-moz-hint'],
  a: ['download'],
  area: ['download', 'alt'],
  // value is special-cased in isAttrNameLocalizable
  input: ['alt', 'placeholder'],
  menuitem: ['label'],
  menu: ['label'],
  optgroup: ['label'],
  option: ['label'],
  track: ['label'],
  img: ['alt'],
  textarea: ['placeholder'],
  th: ['abbr'],
} as Record<string, string[]>

/**
 * Check if attribute is allowed for the given element.
 *
 * This method is used by the sanitizer when the translation markup contains
 * DOM attributes, or when the translation has traits which map to DOM
 * attributes.
 *
 * `explicitlyAllowed` can be passed as a list of attributes explicitly
 * allowed on this element.
 *
 * @param   {string}         name
 * @param   {Element}        element
 * @param   {Array}          explicitlyAllowed
 * @returns {boolean}
 * @private
 */
function isAttrNameLocalizable(
  name: string,
  element: HTMLElement,
  explicitlyAllowed: Array<string> | null = null
): boolean {
  if (explicitlyAllowed && explicitlyAllowed.includes(name)) {
    return true
  }

  if (element.namespaceURI === null) {
    return false
  }

  const attrName = name.toLowerCase()
  const elemName = element.localName

  // Is it a globally safe attribute?
  if (LOCALIZABLE_ATTRIBUTES.global.includes(attrName)) {
    return true
  }

  // Are there no allowed attributes for this element?
  if (!LOCALIZABLE_ATTRIBUTES[elemName]) {
    return false
  }

  // Is it allowed on this element?
  if (LOCALIZABLE_ATTRIBUTES[elemName].includes(attrName)) {
    return true
  }

  // Special case for value on HTML inputs with type button, reset, submit
  if (
    element.namespaceURI === 'http://www.w3.org/1999/xhtml' &&
    elemName === 'input' &&
    attrName === 'value'
  ) {
    const type = (element as HTMLInputElement).type.toLowerCase()
    if (type === 'submit' || type === 'button' || type === 'reset') {
      return true
    }
  }

  return false
}

interface DirectiveBinding {
  arg: string | undefined
  modifiers: Record<string, unknown>
  value: Record<string, any>
}

function translate(el: HTMLElement, fluent: TranslationContext, binding: DirectiveBinding) {
  const key = binding.arg

  if (key === undefined) {
    warn('v-t directive is missing arg with translation key')
    return
  }

  const bundle = fluent.getBundle(key)
  const msg = fluent.getMessage(bundle, key)

  if (bundle === null || msg === null) {
    return
  }

  if (msg.value !== null) {
    el.textContent = fluent.formatPattern(bundle, msg.value, binding.value)
  }

  const allowedAttrs = Object.keys(binding.modifiers)
  for (const [attr, attrValue] of Object.entries(msg.attributes)) {
    if (isAttrNameLocalizable(attr, el, allowedAttrs)) {
      el.setAttribute(attr, fluent.formatPattern(bundle, attrValue, binding.value))
    }
  }
}

export function createVue3Directive(rootContext: TranslationContext) {
  return {
    beforeMount(el: any, binding: any, vnode: any) {
      if (binding.instance == null) {
        return
      }

      const context = getContext(rootContext, binding.instance)
      translate(el, context, binding)
    },

    updated(el: any, binding: any, vnode: any) {
      if (binding.instance == null) {
        return
      }

      const context = getContext(rootContext, binding.instance)
      translate(el, context, binding)
    },
  }
}

export function createVue2Directive(rootContext: TranslationContext) {
  return {
    bind(el: any, binding: any, vnode: any) {
      if (vnode.context == null) {
        return
      }

      const context = getContext(rootContext, vnode.context)
      translate(el, context, binding)
    },

    update(el: any, binding: any, vnode: any) {
      if (vnode.context == null) {
        return
      }

      const context = getContext(rootContext, vnode.context)
      translate(el, context, binding)
    },
  }
}
