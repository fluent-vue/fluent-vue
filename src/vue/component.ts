import { computed, defineComponent, getCurrentInstance, h, inject } from 'vue-demi'
import type { SimpleNode } from 'src/types'
import type { VueComponent } from 'src/types/typesCompat'

import { camelize } from '../util/camelize'
import { getContext } from '../getContext'
import { RootContextSymbol } from '../symbols'
import { assert, warn } from '../util/warn'

function getParentWithFluent(
  instance: VueComponent | null | undefined,
): VueComponent | null | undefined {
  const parent = instance?.$parent
  const target = parent?.$options

  if (target != null && target.fluent == null)
    return getParentWithFluent(parent)

  return parent
}

// Match the opening angle bracket (<) in HTML tags, and HTML entities like
// &amp;, &#0038;, &#x0026;.
const reMarkup = /<|&#?\w+;/

export default defineComponent({
  name: 'i18n',
  props: {
    path: { type: String, required: true },
    tag: { type: String, default: 'span' },
    args: { type: Object, default: () => ({}) },
    html: { type: Boolean, default: false },
  },
  setup(props, { slots, attrs }) {
    const rootContext = inject(RootContextSymbol)
    assert(rootContext != null, 'i18n component used without installing plugin')
    const instance = getCurrentInstance()
    const parent = getParentWithFluent(instance?.proxy)
    const fluent = getContext(rootContext, parent)

    const translation = computed(() => {
      const fluentParams = Object.assign(
        {},
        props.args,
        // Create fake translation parameters for each slot.
        // Later, we'll replace the parameters with the actual slot
        ...Object.keys(slots).map(key => ({
          [key]: `\uFFFF\uFFFE${key}\uFFFF`,
        })),
      )

      const result = fluent.formatWithAttrs(props.path, fluentParams)

      const camelizedAttrs = Object.fromEntries(
        Object.entries(result.attributes).map(([key, value]) => [camelize(key), value]),
      )

      return {
        value: result.value,
        attributes: camelizedAttrs,
      }
    })

    const insertSlots = (text: string | null) => {
      return text?.split('\uFFFF')
        .map(text =>
          text.startsWith('\uFFFE')
            ? slots[text.replace('\uFFFE', '')]!(translation.value.attributes)
            : text,
        )
    }

    // No way to type this properly, so we'll just use `any` for now.
    const processNode = (node: SimpleNode): any => {
      if (node.nodeType === 3) { // Node.TEXT_NODE
        return insertSlots(node.nodeValue)
      }
      else if (node.nodeType === 1) { // Node.ELEMENT_NODE
        const el = node as Element

        return h(
          el.nodeName.toLowerCase(),
          {
            ...Object.fromEntries(
              Array.from(el.attributes).map(attr => [attr.name, attr.value]),
            ),
          },
          Array.from(el.childNodes).map(node => processNode(node)))
      }

      // Ignore other node types for now.
      warn(`Unsupported node type: ${(node as any).nodeType}. If you need support for it, please, create an issue in fluent-vue repository.`)
      return []
    }

    const children = computed(() => {
      // If the message value doesn't contain any markup nor any HTML entities, return it as-is.
      if (!props.html || !reMarkup.test(translation.value.value))
        return insertSlots(translation.value.value)

      // Otherwise, parse the message value as HTML and convert it to an array of VNodes.
      const nodes = fluent.options.parseMarkup(translation.value.value)
      return nodes.map(processNode)
    })

    return () => h(props.tag, { ...attrs }, children.value)
  },
})
