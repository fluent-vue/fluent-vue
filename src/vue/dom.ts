// Copied from fluent-dom library
const LOCALIZABLE_ATTRIBUTES: Record<string, string[]> = {
  global: ['title', 'aria-label', 'aria-valuetext', 'aria-moz-hint'],
  a: ['download'],
  area: ['download', 'alt'],
  // value is special-cased in isAttrNameLocalizable
  input: ['alt', 'placeholder'],
  menuitem: ['label'],
  menu: ['label'],
  optgroup: ['label'],
  option: ['label'],
  track: ['label'],
  img: ['alt'],
  textarea: ['placeholder'],
  th: ['abbr'],
}

/**
 * Check if attribute is allowed for the given element.
 *
 * This method is used by the sanitizer when the translation markup contains
 * DOM attributes, or when the translation has traits which map to DOM
 * attributes.
 *
 * `explicitlyAllowed` can be passed as a list of attributes explicitly
 * allowed on this element.
 *
 * @private
 */
export function isAttrNameLocalizable(
  name: string,
  element: HTMLElement,
  explicitlyAllowed: string[] = [],
): boolean {
  if (explicitlyAllowed.includes(name))
    return true

  const attrName = name.toLowerCase()
  const elemName = element.localName

  // Is it a globally safe attribute?
  if (LOCALIZABLE_ATTRIBUTES.global.includes(attrName))
    return true

  // Are there no allowed attributes for this element?
  if (LOCALIZABLE_ATTRIBUTES[elemName] == null)
    return false

  // Is it allowed on this element?
  if (LOCALIZABLE_ATTRIBUTES[elemName].includes(attrName))
    return true

  // Special case for value on HTML inputs with type button, reset, submit
  if (
    element.namespaceURI === 'http://www.w3.org/1999/xhtml'
    && elemName === 'input'
    && attrName === 'value'
  ) {
    const type = (element as HTMLInputElement).type.toLowerCase()
    if (type === 'submit' || type === 'button' || type === 'reset')
      return true
  }

  return false
}
