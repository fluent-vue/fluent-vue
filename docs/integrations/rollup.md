# Rollup

Rollup support for custom blocks in SFC is provided by [rollup-plugin-fluent-vue](https://www.npmjs.com/package/rollup-plugin-fluent-vue).

```js
import vue from 'rollup-plugin-vue'
import fluentPlugin from 'rollup-plugin-fluent-vue'

module.exports = {
  plugins: [
    vue({
      customBlocks: ['fluent']
    }),
    fluentPlugin()
  ]
};
```

## Options

### blockType

Type: `string`<br>
Default: `fluent`

Custom block tag name.
