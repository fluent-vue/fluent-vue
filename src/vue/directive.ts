import { DirectiveOptions } from 'vue'
import { warn } from '../util/warn'
import { FluentVueObject } from '../interfaces'
import { DirectiveBinding } from 'vue/types/options'

// This part is from fluent-dom library
const LOCALIZABLE_ATTRIBUTES = {
  global: ['title', 'aria-label', 'aria-valuetext', 'aria-moz-hint'],
  a: ['download'],
  area: ['download', 'alt'],
  // value is special-cased in isAttrNameLocalizable
  input: ['alt', 'placeholder'],
  menuitem: ['label'],
  menu: ['label'],
  optgroup: ['label'],
  option: ['label'],
  track: ['label'],
  img: ['alt'],
  textarea: ['placeholder'],
  th: ['abbr']
} as Record<string, string[]>

/**
 * Check if attribute is allowed for the given element.
 *
 * This method is used by the sanitizer when the translation markup contains
 * DOM attributes, or when the translation has traits which map to DOM
 * attributes.
 *
 * `explicitlyAllowed` can be passed as a list of attributes explicitly
 * allowed on this element.
 *
 * @param   {string}         name
 * @param   {Element}        element
 * @param   {Array}          explicitlyAllowed
 * @returns {boolean}
 * @private
 */
function isAttrNameLocalizable(
  name: string,
  element: HTMLElement,
  explicitlyAllowed: Array<string> | null = null
): boolean {
  if (explicitlyAllowed && explicitlyAllowed.includes(name)) {
    return true
  }

  if (element.namespaceURI === null) {
    return false
  }

  const attrName = name.toLowerCase()
  const elemName = element.localName

  // Is it a globally safe attribute?
  if (LOCALIZABLE_ATTRIBUTES.global.includes(attrName)) {
    return true
  }

  // Are there no allowed attributes for this element?
  if (!LOCALIZABLE_ATTRIBUTES[elemName]) {
    return false
  }

  // Is it allowed on this element?
  if (LOCALIZABLE_ATTRIBUTES[elemName].includes(attrName)) {
    return true
  }

  // Special case for value on HTML inputs with type button, reset, submit
  if (
    element.namespaceURI === 'http://www.w3.org/1999/xhtml' &&
    elemName === 'input' &&
    attrName === 'value'
  ) {
    const type = (element as HTMLInputElement).type.toLowerCase()
    if (type === 'submit' || type === 'button' || type === 'reset') {
      return true
    }
  }

  return false
}

function translate(el: HTMLElement, fluent: FluentVueObject, binding: DirectiveBinding) {
  const key = binding.arg

  if (key === undefined) {
    warn('v-t directive is missing arg with translation key')
    return
  }

  const bundle = fluent.getBundle(key)
  const msg = fluent.getMessage(bundle, key)

  if (bundle === null || msg === null) {
    return
  }

  if (msg.value !== null) {
    el.textContent = fluent.formatPattern(bundle, msg.value, binding.value)
  }

  const allowedAttrs = Object.keys(binding.modifiers)
  for (const [attr, attrValue] of Object.entries(msg.attributes)) {
    if (isAttrNameLocalizable(attr, el, allowedAttrs)) {
      el.setAttribute(attr, fluent.formatPattern(bundle, attrValue, binding.value))
    }
  }
}

const directive: DirectiveOptions = {
  bind(el, binding, vnode) {
    if (vnode.context === undefined) {
      return
    }

    translate(el, vnode.context.$fluent, binding)
  },

  update(el, binding, vnode) {
    if (vnode.context === undefined) {
      return
    }

    translate(el, vnode.context.$fluent, binding)
  }
}

export default directive
