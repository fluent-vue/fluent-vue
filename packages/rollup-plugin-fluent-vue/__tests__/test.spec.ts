import { join } from 'path'

import { rollup, RollupOptions } from 'rollup'
import vue from 'rollup-plugin-vue'

import fluentPlugin from '../src'

process.chdir(join(__dirname, 'fixtures'))

const testBundle = async (options: RollupOptions): Promise<string> => {
  const bundle = await rollup({
    ...options,
    external: ['vue', '@fluent/bundle']
  })

  const { output } = await bundle.generate({ format: 'cjs', exports: 'auto' })
  const [{ code }] = output
  return code
}

describe('rollup plugin', () => {
  test('generates custom block code', async () => {
    // Arrange
    // Act
    const code = await testBundle({
      input: 'test.vue',
      plugins: [
        vue({
          customBlocks: ['fluent']
        }),
        fluentPlugin()
      ]
    })

    // Assert
    expect(code).toMatchSnapshot()
  })

  test('custom blockType', async () => {
    // Arrange
    // Act
    const code = await testBundle({
      input: 'blockType.vue',
      plugins: [
        vue({
          customBlocks: ['i18n']
        }), fluentPlugin({
          blockType: 'i18n'
        })
      ],
      external: ['vue', '@fluent/bundle']
    })

    // Assert
    expect(code).toMatchSnapshot()
  })

  test('errors with no locale attr', async () => {
    // Arrange
    const func = async (): Promise<string> => await testBundle({
      input: 'noLocale.vue',
      plugins: [
        vue({
          customBlocks: ['fluent']
        }),
        fluentPlugin()
      ],
      external: ['vue', '@fluent/bundle']
    })

    // Act
    // Assert
    await expect(func).rejects.toThrowErrorMatchingSnapshot()
  })
})
