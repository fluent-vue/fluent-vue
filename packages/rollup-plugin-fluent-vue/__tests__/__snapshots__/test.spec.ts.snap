// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup plugin custom blockType 1`] = `
"'use strict';

var vue = require('vue');
var bundle = require('@fluent/bundle');

function render(_ctx, _cache) {
  return (vue.openBlock(), vue.createBlock(\\"div\\", null, [
    vue.createVNode(\\"div\\", null, vue.toDisplayString(_ctx.$t('hello-user', { userName: _ctx.userName })), 1 /* TEXT */),
    vue.createVNode(\\"div\\", null, vue.toDisplayString(_ctx.$t('shared-photos', { userName: _ctx.userName, photoCount: _ctx.photoCount, userGender: _ctx.userGender })), 1 /* TEXT */)
  ]))
}

function block0 (Component) {
  const target = Component.options || Component;
  target.fluent = target.fluent || {};
  target.fluent['en'] = new bundle.FluentResource(\`
# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.
\`);
}

const script = {};
if (typeof block0 === 'function') block0(script);

script.render = render;
script.__file = \\"blockType.vue\\";

module.exports = script;
"
`;

exports[`rollup plugin errors with no locale attr 1`] = `"Custom block does not have locale attribute"`;

exports[`rollup plugin generates custom block code 1`] = `
"'use strict';

var vue = require('vue');
var bundle = require('@fluent/bundle');

function render(_ctx, _cache) {
  return (vue.openBlock(), vue.createBlock(\\"div\\", null, [
    vue.createVNode(\\"div\\", null, vue.toDisplayString(_ctx.$t('hello-user', { userName: _ctx.userName })), 1 /* TEXT */),
    vue.createVNode(\\"div\\", null, vue.toDisplayString(_ctx.$t('shared-photos', { userName: _ctx.userName, photoCount: _ctx.photoCount, userGender: _ctx.userGender })), 1 /* TEXT */)
  ]))
}

function block0 (Component) {
  const target = Component.options || Component;
  target.fluent = target.fluent || {};
  target.fluent['en'] = new bundle.FluentResource(\`
# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.
\`);
}

const script = {};
if (typeof block0 === 'function') block0(script);

script.render = render;
script.__file = \\"test.vue\\";

module.exports = script;
"
`;
